app/models/fact_value.rb

```diff
@@ -37,12 +37,17 @@ class FactValue < ActiveRecord::Base
   validates :fact_name_id, :uniqueness => { :scope => :host_id }
 
   def self.search_by_host(key, operator, value)
-    search_term = value =~ /\A\d+\Z/ ? 'id' : 'name'
-    conditions = sanitize_sql_for_conditions(["hosts.#{search_term} #{operator} ?", value_to_sql(operator, value)])
-    search     = FactValue.joins(:host).where(conditions).uniq.pluck('fact_values.id')
+    search_by('host', key, operator, value)
+  end
 
-    return { :conditions => "1=0" } if search.empty?
-    { :conditions => "fact_values.id IN(#{search.join(',')})" }
+  def self.search_by_hostgroup(key, operator, value)
+    search_by('hostgroup', key, operator, value)
+  end
+
+  def self.search_by(host_or_hg, key, operator, value)
```

Comment:
you can even DRY this more- key is either `host` or `host.hostgroup`, so you can use the same method for both and just check the key to decide which table to use.